beacon_node/beacon_chain/src/beacon_chain.rs

```diff
@@ -5543,28 +5549,67 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
             let (mut state, state_root) = if let Some((state, state_root)) = head_state_opt {
                 (state, state_root)
             } else {
-                let state_root = head_block.state_root;
-                let state = self
-                    .store
-                    .get_inconsistent_state_for_attestation_verification_only(
-                        &state_root,
-                        Some(head_block.slot),
-                    )?
-                    .ok_or(Error::MissingBeaconState(head_block.state_root))?;
-                (state, state_root)
+                // There's no reason we'd want to load the shuffling for a cold
+                // state; attestation duties will never be pre-finalized and
+                // attestations to finalized blocks are useless.
+                let split_slot = self.store.get_split_slot();
+                if split_slot > target_block.slot {
+                    return Err(Error::RequestedColdShuffling {
+                        split_slot,
+                        request_slot: target_block.slot,
+                    });
+                }
```

Comment:
I'm concerned we have a race condition here that could make this error reachable, resulting in an ugly `error!` [here](https://github.com/sigp/lighthouse/blob/c7e978b727ec7dbc62951af5e92e6d5413e4c74e/beacon_node/network/src/network_beacon_processor/gossip_methods.rs#L2083-L2100). The race is:

1. Verify attestation block isn't finalized (it's not, yet)
2. Check fork choice, attestation block is present
3. Finalization updates, and so does the split point
4. The attestation slot is now prior to the split and triggers this error

Alternatively:

1. Verify attestation block isn't finalized (it's not, yet)
2. Check fork choice, attestation block is present
3. Check attestation slot against split slot, still fine
4. Finalization migration runs and advances the split
5. Try to run `get_state` and fail, because the state has been pruned. _Or_ (today) load a freezer state at great expense ( :scream: ).

This second race is related to #4610.

My preference would be to soak #4160 in Hydra and ensure it's solid before making any changes here. Then we can add this change in maybe with the following defensive changes:

- No new `RequestedColdShuffling` error which complicates error handling (reuse `AttestationStateIsFinalized` which already has its own handling).
- No separate split slot check: use `get_inconsistent_state_for_attestation_verification_only` in both branches and let it do the check atomically (this change exists on the #4160 branch).